### PR TITLE
Adjust preliminary test to remove synthetic test data from train set

### LINKED ISSUE #7: Train/test splits in `example.py` script have test data bleeding into training set
Since rebalancing happens on _all_ the data, this portion of code that splits training and test sets is letting in variations of test data into the training set. This is because it sets aside all synthetic data before splitting into train/test sets, then it appends all synthetic data to the training set.

https://github.com/byu-dml/d3m-profiler/blob/9a3bc45061267091b0109f2159648785e370a18b/example.py#L63-L75

Perhaps a more accurate way of splitting and balancing data would be to do it in the following order:
1. Split organic data into train/test sets
1. Balance training data
1. Train model and get predictions

### PR DESCRIPTION
**Pull request by e13h**
Refactored `example.py` to make it a little easier to follow.

Now, when rebalancing the data, it only rebalances the training set (not the whole data frame). It also uses the exact same train/test split as when fitting unbalanced data, so there is a more uniform comparison.

Closes #7

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,3 +21,12 @@ Run models with `python3 evaluate_models.py [<number of cores to use>]`
...
 
 Score predictions with `python3 score_results.py results[_small]/<filename>.csv`
+
+## Creating a `conda` environment
+To create a `conda` virtual environment for running `example.py`, run the following commands:
+```
+conda create --name example cudatoolkit=9.0 tensorflow tensorflow-gpu cycler joblib kiwisolver matplotlib numpy pandas pyparsing python-dateutil pytz scikit-learn scipy six
+
+conda activate example
+pip install imbalanced-learn==0.6.2 sent2vec-prebuilt==0.0.3
```
Comment (bjschoenfeld) [inline README.md]: Can conda use the dependencies listed in `requirements.txt`?

File: d3m_profiler/rebalance.py [hunk 2]
```diff
@@ -137,15 +137,18 @@ def _construct_rebalanced_df(X_resampled: pd.DataFrame, Y_resampled: pd.DataFram
...
     The rebalanced DataFrame 
 """
-def rebalance_SMOTE(df: pd.DataFrame, type_column: str, method: str, model_weights_path: str) -> pd.DataFrame:
+def embed_and_rebalance_SMOTE(df: pd.DataFrame, type_column: str, method: str, model_weights_path: str) -> pd.DataFrame:
     embedded_df = embed(df, type_column, model_weights_path)
 
+    return rebalance_SMOTE(embedded_df, type_column, method)
+    
+
+def rebalance_SMOTE(embedded_df: pd.DataFrame, type_column: str, method: str) -> pd.DataFrame:
```
Comment (bjschoenfeld) [inline d3m_profiler/rebalance.py]: Perhaps we should decouple the embedding and the rebalancing functions.

File: example.py [hunk 3]
```diff
@@ -31,74 +26,60 @@
 type_column = 'colType'
 model_weights_path = 'torontobooks_unigrams.bin'
 
-open_d3m_file = 'data/open_d3m_data.csv'
-closed_d3m_file = 'data/closed_d3m_data.csv'
+open_d3m_file = '~/data/open_d3m_unembed_data.csv'
+closed_d3m_file = '~/data/closed_d3m_unembed_data.csv'
```
Comment (bjschoenfeld) [inline example.py]: can we add the data to this repo? maybe a compressed version.

Comment (bjschoenfeld) [reply example.py]: can we also rename `open` to `public` and `closed` to `private`

File: example.py [hunk 4]
```diff
@@ -31,74 +26,60 @@
...
+    f1_micro = f1_score(ytest, yhat, average='micro')
+    f1_macro = f1_score(ytest, yhat, average='macro')
+    f1_weighted = f1_score(ytest, yhat, average='weighted')
+
+    return {'data_collection': data_collection, 'classifier': model_class.__name__, 'balanced': isBalanced, 'accuracy_score': accuracy, 'f1_score_micro': f1_micro, 'f1_score_macro': f1_macro, 'f1_score_weighted': f1_weighted}
+
+def split_X_and_y(data: pd.DataFrame):
+    X = data.drop(['datasetName', type_column], axis=1)
+    y = data[type_column]
+    return X, y
```
Comment (bjschoenfeld) [inline example.py]: the name `split` has a connotation of train/test split. is there another name we could use? `extract`? `separate`?

File: README.md (hunk 1 above)
Comment (e13h) [reply README.md]: I ran into some errors because there are dependencies in our `requirements.txt` that have a version of `0.0.0` for some reason. `pip` handles this as "oh, that dependency is met" while `conda` says "oh no, can't find that version"
https://github.com/byu-dml/d3m-profiler/blob/9a3bc45061267091b0109f2159648785e370a18b/requirements.txt#L8
https://github.com/byu-dml/d3m-profiler/blob/9a3bc45061267091b0109f2159648785e370a18b/requirements.txt#L16

File: example.py (hunk 3 above)
Comment (e13h) [reply example.py]: @bjschoenfeld is it okay if the metadata for the private database (`closed_d3m_unembed_data.csv` above) is hosted on our repo? It's not _the_ data, but it includes all the private dataset names and descriptions.

Comment (e13h) [reply example.py]: @bjschoenfeld I uploaded a compressed version of the public metadata `.csv` file. Let me know if I have the go-ahead to do the same for the private metadata `.csv` file too.

File: README.md (hunk 1 above)
Comment (bjschoenfeld) [reply README.md]: Ah yes, I have seen this issue before. Both of those dependencies should be removed from `requirements.txt`. Can you remove them and then update the readme to use the requirements.txt?

File: example.py (hunk 3 above)
Comment (bjschoenfeld) [reply example.py]: maybe to be safe, let's not publish the private one.

**Comment by e13h** [comment]
Closed because of staleness
